gateway-ha/src/main/java/io/trino/gateway/ha/router/ExternalRoutingGroupSelector.java

```diff
@@ -0,0 +1,153 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpClientConfig;
+import io.airlift.http.client.JsonBodyGenerator;
+import io.airlift.http.client.JsonResponseHandler;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.jetty.JettyHttpClient;
+import io.airlift.json.JsonCodec;
+import io.airlift.log.Logger;
+import io.trino.gateway.ha.config.RequestAnalyzerConfig;
+import io.trino.gateway.ha.config.RulesExternalConfiguration;
+import io.trino.gateway.ha.router.schema.RoutingGroupExternalBody;
+import io.trino.gateway.ha.router.schema.RoutingGroupExternalResponse;
+import jakarta.servlet.http.HttpServletRequest;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static io.airlift.http.client.JsonBodyGenerator.jsonBodyGenerator;
+import static io.airlift.http.client.JsonResponseHandler.createJsonResponseHandler;
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static java.util.Collections.list;
+import static java.util.Objects.requireNonNull;
+
+public class ExternalRoutingGroupSelector
+        implements RoutingGroupSelector
+{
+    private static final Logger log = Logger.get(ExternalRoutingGroupSelector.class);
+    private final Set<String> blacklistHeaders;
+    private final URI uri;
+    private final HttpClient httpClient;
+    private final RequestAnalyzerConfig requestAnalyzerConfig;
+    private final TrinoRequestUser.TrinoRequestUserProvider trinoRequestUserProvider;
+    private static final JsonCodec<RoutingGroupExternalBody> ROUTING_GROUP_EXTERNAL_BODY_JSON_CODEC = jsonCodec(RoutingGroupExternalBody.class);
+    private static final JsonResponseHandler<RoutingGroupExternalResponse> ROUTING_GROUP_EXTERNAL_RESPONSE_JSON_RESPONSE_HANDLER =
+            createJsonResponseHandler(jsonCodec(RoutingGroupExternalResponse.class));
+
+    @VisibleForTesting
+    ExternalRoutingGroupSelector(RulesExternalConfiguration rulesExternalConfiguration, RequestAnalyzerConfig requestAnalyzerConfig)
+    {
+        Set<String> defaultBlacklistHeaders = ImmutableSet.of("Content-Length");
+        this.blacklistHeaders = ImmutableSet.<String>builder()
+                .addAll(defaultBlacklistHeaders)
+                .addAll(rulesExternalConfiguration.getBlackListHeaders())
+                .build();
+
+        this.requestAnalyzerConfig = requestAnalyzerConfig;
+        trinoRequestUserProvider = new TrinoRequestUser.TrinoRequestUserProvider(requestAnalyzerConfig);
+        try {
+            this.uri = new URI(requireNonNull(rulesExternalConfiguration.getUrlPath(),
+                    "Invalid url provided, using routing group header as default."));
```

Comment:
```suggestion
                    "Invalid URL provided, using routing group header as default."));
```

or use the parameter name

Invalid URL provided in urlPath ..